src/snippet/mod.rs

```diff
@@ -0,0 +1,207 @@
+    let mut token_stream = tokenizer.token_stream(text);
+    let mut fragment = FragmentCandidate::new(0, 0);
+    let mut fragments:Vec<FragmentCandidate> = vec![];
+
+    loop {
+        if let Some(next) = token_stream.next() {
+            if (next.offset_to - fragment.start_offset) > max_num_chars {
+                let txt = &text[fragment.start_offset..fragment.stop_offset];
+                if fragment.score > 0.0 {
+                    fragments.push(fragment)
+                };
+                fragment = FragmentCandidate::new(next.offset_from, next.offset_to);
+            } else {
+                fragment.calculate_score(next, &terms);
+                fragment.stop_offset = next.offset_to;
+            }
+        } else {
+            let txt = &text[fragment.start_offset..fragment.stop_offset];
+            if fragment.score > 0.0 {
+                fragments.push(fragment)
+            };
+            break;
+        }
+    }
+
+    fragments
+}
+
+/// Returns a Snippet
+///
+/// Takes a vector of `FragmentCandidate`s and the text.
+/// Figures out the best fragment from it and creates a snippet.
+fn select_best_fragment_combination<'a>(fragments: Vec<FragmentCandidate>,
+                                        text: &'a str,) -> Snippet {
+    if let Some(init) = fragments.iter().nth(0) {
+        let fragment = fragments.iter().skip(1).fold(init, |acc, item| {
+            if item.score > acc.score { item } else { acc }
+        });
+        let fragment_text = &text[fragment.start_offset..fragment.stop_offset];
+        let highlighted = fragment.highlighted.iter().map(|item| {
+            HighlightSection{start: item.start-fragment.start_offset,
+                             stop: item.stop-fragment.start_offset}
+        }).collect();
+        Snippet{fragments: fragment_text.to_owned(),
+                highlighted: highlighted}
+    } else {
+        // when there no fragments to chose from,
+        // for now create a empty snippet
+        Snippet{fragments: String::new(),
+                highlighted: vec![]}
+    }
+}
+
+pub fn generate_snippet<'a>(
+    doc: &'a [FieldValue],
+    index: &Index,
+    terms: Vec<Term>,
+    max_num_chars: usize) -> Snippet {
+    unimplemented!();
+}
+
+
+#[cfg(test)]
+mod tests {
+    use tokenizer::{SimpleTokenizer, box_tokenizer};
+    use std::iter::Iterator;
+    use std::collections::BTreeMap;
+    use super::{search_fragments, select_best_fragment_combination};
+
+    #[test]
+    fn test_snippet() {
+        let tokenizer = SimpleTokenizer;
+
+        let t = box_tokenizer(tokenizer);
+
+        let text = "Rust is a systems programming language sponsored by Mozilla which describes it as a \"safe, concurrent, practical language\", supporting functional and imperative-procedural paradigms. Rust is syntactically similar to C++[according to whom?], but its designers intend it to provide better memory safety while still maintaining performance.
```

Comment:
That's a good unit test. Keep it. Can you add more simpler unit test (the words can be one letter words) that explore the specification.

For instance :
- What happens if the first token is a keyword?
- if the text is empty?
- if the text contains no keyword?